feat: add exporting state to dynamic configuration

lenaschoenburg opened the pull request:
This adds a new field to dynamic partition configuration that controls the exporting state: Exporting, Soft Paused or Paused.

For now, this has no function, but it's at least (de-)serialized to the protobuf representation and thus persisted. When this field is missing from the serialized state, for example after update from a previous version, we will use `null` as a placeholder and re-encode it as `EXPORTING_STATE_UNKNOWN`. We also use `null` when initializing dynamic configuration for the first time. Eventually, this will be initialized based on the actual state of exporting as it was set by `PartitionAdminControl`.

I've included a couple of other refactors and fixes that are related and added the motivation in each commit message.

relates to #39743

lenaschoenburg committed:
docs: update `DynamicPartitionConfig` based on actual usage

Both `init` and `uninitialized` is actually used in production code now
and not just a temporary workaround.

`ExporterStateInitializer` uses `init` to initially create and then sync
the state with static configuration.

`uninitialized` is usd by the protobuf decoder for backwards
compatibility.

lenaschoenburg committed:
feat: add exporting state to dynamic partition config

For now, this has no function, but it's at least (de-)serialized to
the protobuf representation. When this field is missing from the
serialized state, we will use `null` as a placeholder and re-encode
it as `EXPORTING_STATE_UNKNOWN`. We also use `null` when initializing
dynamic configuration for the first time. Eventually, this will be
initialized based on the actual state of exporting as it was set by
`PartitionAdminControl`.

lenaschoenburg committed:
refactor: defensive copies in `ClusterConfiguration` and nested types

This is generally a good practice because it removes one potential
source of mutability for records which are assumed to be immutable.

This was motivated by slow property tests. So far, we had to rely on
AssertJ's recursive comparison to compare two `ClusterConfiguration`
objects. This was caused by jqwik interfering by using sorted collection
types when creating `ClusterConfiguration` instances.
On deserialization, we would instead use unsorted collection types,
leading to a failed `equals` check. With defensive copies, we discard
jqwik's sorted collections and always use unsorted ones, letting the
`equals` check pass again. A regular `equals` is much faster than
recursive comparison, reducing the overall run time of
`ProtoBufSerializerPropertyTest` from ~20 seconds to ~6 seconds.

lenaschoenburg committed:
refactor: partition members that require reconfiguration must be unique

We know that partition members are always unique (one broker can't
replicate a single partition twice), so we can narrow the type from
`Collection` to `Set`.

lenaschoenburg committed:
test: cover missing exporting state in property tests

lenaschoenburg committed:
test: verify that exporting config without status is decoded correctly

lenaschoenburg committed:
fix: configuration change requests generate operations deterministically

The order of entries in the members and partitions maps is undefined,
leading to a non-deterministic order of operations.
This doesn't cause any big issues but is not ideal, especially because
it could differ to previous dry-run results.

Additionally, this caused tests that rely on the exact order to fail,
now that we take defensive copies of these collections.

entangled90 reviewed (commented):
Thanks!
I like the idea of making defensive copies to avoid very difficult to understand bugs.
As we discussed privately we should:
- Set all collections type to be *ordered* inside the record
- Create constructors that take *unordered* collections and make a copy with an *ordered* collection
- Collections should be immutable (aka persistent) or effectively immutable 

In this way, iterating over the collection will always result in a predefined order as it's so easy to forget a `sorted()` before iterating.

lenaschoenburg committed:
feat: validate `ClusterConfiguration` components

lenaschoenburg committed:
feat: fully sort `ClusterConfiguration` for determinism

This makes it easier for consumers of `ClusterConfiguration` to produce
data deterministically. For example, this makes the Cluster API
responses more predictable and easier to compare.

lenaschoenburg committed:
feat: fully sort `ClusterConfigurationChangeResponse` for determinism

Combined with 7757b2a7ecdc0a5959e8b0e89a44263e869090ef, this should make
the Cluster API responses fully deterministic

lenaschoenburg committed:
test: provide arbitrary sorted sets and maps

Now that we use those types in `ClusterConfiguration`, we need to tell
jqwik how to generate these types or instruct it to use the constructors
which take unsorted maps and sets and converts them to sorted ones
automatically.

Having a generic provider for all sorted maps and sets seemed like
the easier and more scalable approach.

lenaschoenburg committed:
refactor: handle unknown exporting state explicitly

entangled90 commented on zeebe/dynamic-config/src/main/java/io/camunda/zeebe/dynamic/config/api/ForceScaleDownRequestTransformer.java:
```diff
@@ -118,8 +121,9 @@ private List<ClusterConfigurationChangeOperation> reconfigurePartitions(
 
   private List<ClusterConfigurationChangeOperation> forceRemoveMembers(
-      final List<MemberId> membersToRemove) {
+      final Set<MemberId> membersToRemove) {
```
This is a private function, so we can be more explicit?

lenaschoenburg replied:
Implemented in 128db19b4baac23961dd5089adc625b6e26cf960

entangled90 commented on zeebe/dynamic-config/src/main/java/io/camunda/zeebe/dynamic/config/state/ClusterConfigurationChangeOperation.java:
```diff
@@ -89,7 +95,11 @@ record AwaitRedistributionCompletion(
-        implements ScaleUpOperation {}
+        implements ScaleUpOperation {
+      public AwaitRelocationCompletion {
+        partitionsToRelocate = new TreeSet<>(partitionsToRelocate);
```
same as above

entangled90 commented on zeebe/dynamic-config/src/test/java/io/camunda/zeebe/dynamic/config/PersistedClusterConfigurationRandomizedPropertyTest.java:
```diff
@@ -42,7 +42,6 @@ void shouldUpdatePersistedFile(
     // then
     assertThat(updatedTopology).isEqualTo(persistedClusterTopology.getConfiguration());
     assertThat(PersistedClusterConfiguration.ofFile(topologyFile, serializer).getConfiguration())
-        .usingRecursiveComparison()
```
:1st_place_medal:

entangled90 reviewed (approved):
In hindsight we should have split this PR into 2 different ones, one with just the refactor and one with the extra field. 

Anyway, looks good. Pre-approving, just take a look at the comments

lenaschoenburg commented on zeebe/dynamic-config/src/main/java/io/camunda/zeebe/dynamic/config/state/ClusterConfigurationChangeOperation.java:
```diff
@@ -80,7 +82,11 @@ record StartPartitionScaleUp(MemberId memberId, int desiredPartitionCount)
-        implements ScaleUpOperation {}
+        implements ScaleUpOperation {
+      public AwaitRedistributionCompletion {
+        partitionsToRedistribute = new TreeSet<>(partitionsToRedistribute);
```
Amended in 9549e46c73fbfa7ab95c4da42676cb295b86569c
